v1/services/storage/store.go

```diff
@@ -239,6 +272,17 @@ func (s *Store) TagKeys(ctx context.Context, req *datatypes.TagKeysRequest) (cur
 		if found := reads.HasFieldValueKey(expr); found {
 			return nil, errors.New("field values unsupported")
 		}
+		if found := reads.ExprHasKey(expr, fieldKey); found {
+			mqAttrs := &metaqueryAttributes{
+				orgID: source.GetOrgID(),
+				db:    db,
+				rp:    rp,
+				start: start,
+				end:   end,
+				pred:  expr,
+			}
+			return s.tagKeysWithFieldPredicate(ctx, mqAttrs, shardIDs)
+		}
 		// this will remove any _field references, which are not indexed
 		//   see https://github.com/influxdata/influxdb/issues/19488
 		expr = influxql.Reduce(RewriteExprRemoveFieldKeyAndValue(influxql.CloneExpr(expr)), nil)
```

Comment:
This logic here is no longer needed, since we won't have any predicate on `$` since we would have returned an error above, nor `_field`, since we would have taken the slow path.

Reply:
It's still worth calling `Reduce(influxql.CloneExpr(expr))` though since something like `(r) => true` might have gotten pushed down.